### PR TITLE
Create a levelbuilder-facing table of all code docs

### PR DESCRIPTION
**Pull request by bethanyaconnor**
Second to last PR for the all code docs page (well, until I have to update it for Java). The code change to add this page (even with two PRs already split out) is over 1000 lines so I decided to split out the component that loads a table of code docs (this PR) as it's more complicated than the others. Apologies that this is still a large PR on its own.

This component loads a table of code docs, optionally filtered using dropdown selections. I chose 20 entries per page a bit arbitrarily but it feels good for the page. These docs then have a button that opens a new tab to edit, a button to clone to another programming environment (dialog added in #45285). The routes to clone and destroy a programming expression were added in #45174.`get_filtered_expressions`, which gets the programming expressions according to the filters, is a new route added in this PR.

Changing environment and page:

https://user-images.githubusercontent.com/46464143/159547668-9313997f-46ac-4cda-93e9-e2c11c673190.mov

Changing category:

https://user-images.githubusercontent.com/46464143/159547771-e84a161d-ef50-42b0-bc5c-c9a8b403d29f.mov

Deleting a programming expression:

https://user-images.githubusercontent.com/46464143/159547848-810b00ec-feca-4670-bb65-e103527f391a.mov

## Links

[jira](https://codedotorg.atlassian.net/browse/PLAT-695)
[product spec](https://docs.google.com/document/d/1CuMG8vlECM0xLs8kq5hocUS5ToyPWipjTGU841W2wvs/edit#bookmark=id.dwxvqefs3lom)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/controllers/programming_expressions_controller.rb [hunk 1]
```diff
@@ -1,8 +1,24 @@
...
 
+  # GET /programming_expressions/get_filtered_expressions
+  def get_filtered_expressions
+    @programming_expressions = ProgrammingExpression.all
+    @programming_expressions = @programming_expressions.where(programming_environment_id: params[:programmingEnvironmentId]) if params[:programmingEnvironmentId]
+    @programming_expressions = @programming_expressions.where(programming_environment_category_id: params[:categoryId]) if params[:categoryId]
+
+    results_per_page = 20
+    total_expressions = @programming_expressions.length
+    num_pages = params[:page].blank? ? 1 : (total_expressions / results_per_page.to_f).ceil
```
Comment (jamescodeorg) [inline dashboard/app/controllers/programming_expressions_controller.rb]: Mostly curious -- is there a scenario where we're going to call this API without the page parameter?  It seems like it might be safer (and simpler) to enforce that this parameter is present?

File: dashboard/app/controllers/programming_expressions_controller.rb [hunk 2]
```diff
@@ -1,8 +1,24 @@
 class ProgrammingExpressionsController < ApplicationController
+  include Rails.application.routes.url_helpers
+
   load_and_authorize_resource
 
   before_action :require_levelbuilder_mode_or_test_env, except: [:search, :show, :show_by_keys]
 
+  # GET /programming_expressions/get_filtered_expressions
```
Comment (jamescodeorg) [inline dashboard/app/controllers/programming_expressions_controller.rb:8]: It might be worth listing the params that are accepted here and noting that page is 1-based (right?).

File: apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+          text=""
+          icon="trash"
+          color="red"
+          style={{margin: 0}}
+        />
+      </div>
+    );
+  };
+
+  const fetchExpressions = (environmentId, categoryId, page) => {
```
Comment (jamescodeorg) [inline apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx:66]: I think the useFetch hook that I added recently could be useful here (but also ok if you don't want to use it).  (I haven't shared it broadly yet because I wanted to battle test it in the Lesson Lock dialog first but that work keeps getting pushed back.)

File: apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+      .then(data => {
+        if (success) {
+          setProgrammingExpressions(data.expressions);
+          setNumPages(data.numPages);
+        }
+      });
+  };
+
+  const destroyExpression = (destroyPath, callback) => {
+    fetch(destroyPath, {
```
Comment (jamescodeorg) [inline apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx:100]: In some other places where we use fetch, we've needed to add a `credentials: same-origin` option to make sure the session cookie is sent.  Is the back-end getting a valid current_user in this codepath?

File: apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+import * as Table from 'reactabular-table';
+import queryString from 'query-string';
+import Button from '@cdo/apps/templates/Button';
+import StylizedBaseDialog from '@cdo/apps/componentLibrary/StylizedBaseDialog';
+import PaginationWrapper from '@cdo/apps/templates/PaginationWrapper';
+import CloneProgrammingExpressionDialog from './CloneProgrammingExpressionDialog';
+
+const DEFAULT_VALUE = 'all';
+
+export default function ProgrammingExpressionsTable({
```
Comment (jamescodeorg) [inline apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx:18]: I think it would be useful to have a header comment that gives a description of the component to help orient someone who might not be familiar with the code.

File: apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+  programmingEnvironmentsForSelect,
+  categoriesForSelect,
```
Comment (jamescodeorg) [inline apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx]: This might just be me, but I found these names a little confusing (since I didn't realize "select" referred to a select tag).  I'm wondering if something like `allProgrammingEnvironments` and `allCategories` would be clearer?  Up to you.

File: apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+      setCategoriesAvailableForSelect(
+        categoriesForSelect.filter(
+          cat => String(cat.envId) === String(newSelectedEnvironment)
+        )
+      );
+      if (String(selectedCategory.envId) !== newSelectedEnvironment) {
+        setSelectedCategory(DEFAULT_VALUE);
+      }
+    }
+    setCurrentPage(1);
```
Comment (jamescodeorg) [inline apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx:132]: (minor) Don't know if this would be better or worse, but I think you can determine when to call `setCurrentPage` using useEffect() with a dependency on `selectedCategory` and `selectedEnvironment`.

Actually, I think you can express the relationship between selectedEnvironment, selectedCategory, and page all through useEffect.  I'm new to this style of writing components though and haven't decided if this is amazing (because it's a lot more declarative) or really hard to follow.

File: apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+                  selectedEnvironment,
+                  selectedCategory,
+                  currentPage
+                );
+              }
+            );
+          }}
+          handleClose={() => setItemToDelete(null)}
+          isOpen
+        />
```
Comment (jamescodeorg) [inline apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx]: (minor) Maybe split this out in a separate function called `renderDeleteConfirmationDialog` or something like that?  Same comment for the other dialog below.

File: apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+        }}
+        value={selectedCategory}
+      >
+        <option value={DEFAULT_VALUE}>All Categories</option>
+        {categoriesAvailableForSelect.map(category => (
+          <option key={category.id} value={category.id}>
+            {category.formattedName}
+          </option>
+        ))}
+      </select>
```
Comment (jamescodeorg) [inline apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx]: (minor) Maybe split this out into a `renderFilters` function?

**Review by jamescodeorg — COMMENTED** [review]
Generally looks great!  I found it a bit hard to parse the ProgrammingExpressionsTable component at first, maybe some extra comments or splitting up the main render method might help a bit?

(It kind of bugs me that our javascript style is so "vertical" but that's a rant for another day.)

File: apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx (hunk 3 above)
Comment (bethanyaconnor) [reply apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx:66]: I saw that but I wasn't quite sure how to use it. I had already started working on this when you created that hook and I really want to get this in, but I'm happy to collaborate as a follow-up on using that hook!

File: dashboard/app/controllers/programming_expressions_controller.rb (hunk 2 above)
Comment (bethanyaconnor) [reply dashboard/app/controllers/programming_expressions_controller.rb:8]: Yep, it's 1 indexed. Based on my quick experiments, that's how Rails/SQL works but I couldn't find any documentation on that...

File: dashboard/app/controllers/programming_expressions_controller.rb (hunk 1 above)
Comment (bethanyaconnor) [reply dashboard/app/controllers/programming_expressions_controller.rb]: I can't think of one so if that is simpler, I can enforce it!

File: apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx (hunk 4 above)
Comment (bethanyaconnor) [reply apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx:100]: Yes it is. It looks like [same-origin is the default here](https://github.com/whatwg/fetch/pull/585)

File: apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx (hunk 8 above)
Comment (bethanyaconnor) [reply apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx]: I ended up splitting both of these into one function `renderDialogs()`. Is that okay would you rather they be their own functions?

Side note: the render function looks a lot cleaner with the various render functions pulled out :)

File: apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx (hunk 7 above)
Comment (bethanyaconnor) [reply apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx:132]: To clarify, you're suggesting adding another `useEffect` that is:
```
useEffect(() => {
  setPage(1);
}, [selectedEnvironment, selectedCategory])
```
I think that would work but I'm having trouble justifying why that would be clearer than just doing it in the handler. Plus React handles state updates better in a click handler (related [bug](https://github.com/facebook/react/issues/14259)).

I might be misunderstanding your suggestion though

File: apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx (hunk 8 above)
Comment (jamescodeorg) [reply apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx]: That's fine, thanks for trying it out!

File: apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx (hunk 7 above)
Comment (jamescodeorg) [reply apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx:132]: Yes, that was what I was suggesting, but maybe that's not better.  The other thing that comes to mind is that these three pieces of state are actually related so maybe something like `useReducer` would model that more explicitly.  But also ok to leave as-is, up to you.

Comment (bethanyaconnor) [reply apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx:132]: @jamescodeorg I tried using a reducer in df837a06f79 and I'm curious what you think. On one hand, it does move all the logic around the four entangled pieces of state into one place, which is nice. On the other hand, it didn't make sense to me to move all the state to the reducer so now the state is handled in two different ways. I could bring the rest of the state up into the reducer but that state is very simple.

I haven't used `useReducer` before and it doesn't look like it's been used in our codebase so I'm not sure what the conventions here are. It was easy to do (and a good leaning opportunity!) so if it isn't better, I'll just rebase it out.

File: apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+      queryString.stringify(data);
+    let success = false;
+    fetch(url)
+      .then(response => {
+        if (response.ok) {
+          success = true;
+        } else {
+          setError(response.statusText);
+        }
+        return response.json();
```
Comment (jamescodeorg) [inline apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx:86]: What happens if response.json() fails here (e.g. if the response body is not vaild json)?

I'm wondering if there are error cases in additional to the one handled on line 124 and if/how they should be handled.

**Review by jamescodeorg — COMMENTED** [review]
> @jamescodeorg I tried using a reducer in https://github.com/code-dot-org/code-dot-org/commit/df837a06f7926a7db4cbcc361e7d1a0296bf05fc and I'm curious what you think. On one hand, it does move all the logic around the four entangled pieces of state into one place, which is nice. On the other hand, it didn't make sense to me to move all the state to the reducer so now the state is handled in two different ways. I could bring the rest of the state up into the reducer but that state is very simple.

> I haven't used useReducer before and it doesn't look like it's been used in our codebase so I'm not sure what the conventions here are. It was easy to do (and a good leaning opportunity!) so if it isn't better, I'll just rebase it out.

Thanks for trying this out!  Now that I see it, I think the previous code (**not** using `useReducer`) was a bit easier for me to follow but I am ok with either approach.

**Comment by bethanyaconnor** [comment]
> Thanks for trying this out! Now that I see it, I think the previous code (not using useReducer) was a bit easier for me to follow but I am ok with either approach.

I think I agree, plus I'd prefer to keep this file consistent so I'll take it out.

File: apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx (hunk 10 above)
Comment (bethanyaconnor) [reply apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionsTable.jsx:86]: Good question. I'll add error handling.

**Review by jamescodeorg — APPROVED** [review]
Thanks for making all of the changes!
